File: lophine-server/src/main/java/fun/bm/mili/command/counter/CounterCommand.java
package fun.bm.mili.command.counter;

import com.mojang.brigadier.exceptions.CommandSyntaxException;
import fun.bm.mili.command.counter.sub.DisplayCommand;
import fun.bm.mili.command.counter.sub.ResetCommand;
import fun.bm.mili.command.counter.sub.ToggleCommand;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.JoinConfiguration;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.command.CommandSender;
import org.jetbrains.annotations.NotNull;
import org.leavesmc.leaves.command.CommandContext;
import org.leavesmc.leaves.command.RootNode;
import org.leavesmc.leaves.util.HopperCounter;

public class CounterCommand extends RootNode {
    private final String PERM_BASE;

    public CounterCommand() {
        super("counter", "lophine.commands.counter");
        this.PERM_BASE = "lophine.commands.counter";
        children(
                new ToggleCommand(this),
                new ResetCommand(this),
                new DisplayCommand(this)
        );
    }

    @Override
    protected boolean execute(@NotNull CommandContext context) throws CommandSyntaxException {
        context.getSender().sendMessage(Component.join(JoinConfiguration.noSeparators(),
                Component.text("Hopper Counter: ", NamedTextColor.GRAY),
                Component.text(HopperCounter.isEnabled(), HopperCounter.isEnabled() ? NamedTextColor.AQUA : NamedTextColor.GRAY)
        ));
        return true;
    }

    public boolean hasPermission(@NotNull CommandSender sender, String... subcommand) {
        return hasPermission(PERM_BASE, sender, subcommand);
    }
}

File: mili-server-kotlin/src/main/kotlin/fun/bm/mili/command/counter/CounterCommand.kt
package `fun`.bm.mili.command.counter

import `fun`.bm.mili.command.counter.sub.DisplayCommand
import `fun`.bm.mili.command.counter.sub.ResetCommand
import `fun`.bm.mili.command.counter.sub.ToggleCommand
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.JoinConfiguration
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.command.CommandSender
import org.leavesmc.leaves.command.CommandContext
import org.leavesmc.leaves.command.RootNode
import org.leavesmc.leaves.util.HopperCounter

class CounterCommand : RootNode("counter", "mili.commands.counter") {

    private val permBase = "mili.commands.counter"

    init {
        children(
            ::ToggleCommand,
            ::ResetCommand,
            ::DisplayCommand,
        )
    }

    override fun execute(context: CommandContext): Boolean {
        context.sender.sendMessage(
            Component.join(
                JoinConfiguration.noSeparators(),
                Component.text("Hopper Counter: ", NamedTextColor.GRAY),
                Component.text(
                    HopperCounter.isEnabled(),
                    if (HopperCounter.isEnabled()) NamedTextColor.AQUA else NamedTextColor.GRAY,
                ),
            ),
        )
        return true
    }

    fun hasPermission(sender: CommandSender, vararg subcommand: String): Boolean {
        val suffix = subcommand.joinToString(".")
        return if (suffix.isEmpty()) {
            sender.hasPermission(permBase)
        } else {
            sender.hasPermission(permBase) || sender.hasPermission("$permBase.$suffix")
        }
    }
}